### PR TITLE
Add leniency to control message misbehaviors on cluster prefixed topics 

### PR DESCRIPTION
**Pull request by kc1116**
This PR introduces the `clusterPrefixedPenaltyReductionFactor`, used to add leniency to scoring penalties for misbehaviors on cluster-prefixed topics. Nodes can occasionally fall behind in protocol state, leading to requests and transmissions of outdated information. This situation can result in nodes being penalized and relegated to the network's periphery, which in turn extends the time required for these nodes to catch up with the rest of the network. This can be an issue on cluster prefixed topics that are crucial to the move the chain along ie: consensus cluster, collections cluster.

The clusterPrefixedPenaltyReductionFactor allows nodes that participate in cluster communication and have fallen behind to be penalized to an extent where they do not actively participate in message routing, but still enables them to catch up within an acceptable timeframe. This strategic addition ensures that network performance remains robust even when nodes experience temporary setbacks.

### CODE REVIEW

File: network/p2p/inspector/validation/control_message_validation_inspector.go
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:1]: Please implement inspector tests assessing that when an RPC contains cluster-prefix violations, the notifications that are sent as the result have `IsClusterPrefix` value as true. The test can be implemented by refactoring the existing tests running in two modes of cluster and non-cluster prefixes.

File: network/p2p/consumers.go [hunk 1]
```diff
@@ -33,14 +33,17 @@ type InvCtrlMsgNotif struct {
 	Error error
 	// MsgType the control message type.
 	MsgType p2pmsg.ControlMessageType
+	// IsClusterPrefixed indicates if the error occurred on a cluster prefixed topic of the control message.
+	IsClusterPrefixed bool
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/consumers.go:43]: https://github.com/onflow/flow-go/pull/4857/commits/f512095ff6cb88cabb5718666715dabe6d504b8e

File: network/p2p/scoring/registry.go [hunk 2]
```diff
@@ -268,6 +275,12 @@ func (r *GossipSubAppSpecificScoreRegistry) OnInvalidControlMessageNotification(
 			// the error is considered fatal as it means that we have an unsupported misbehaviour type, we should crash the node to prevent routing attack vulnerability.
 			lg.Fatal().Str("misbehavior_type", notification.MsgType.String()).Msg("unknown misbehaviour type")
 		}
+
+		// reduce penalty for cluster prefixed topics allowing nodes that are potentially behind to catch up
+		if notification.IsClusterPrefixed {
+			record.Penalty *= r.penalty.ClusterPrefixedPenaltyReductionFactor
```
[the remark this answers is not in the document]
Comment (kc1116) [reply network/p2p/scoring/registry.go]: https://github.com/onflow/flow-go/pull/4857/commits/ae59eb23581f560eb06c6ed85ec20d38f208a6e3

File: network/p2p/inspector/validation/control_message_validation_inspector.go
Comment (kc1116) [reply network/p2p/inspector/validation/control_message_validation_inspector.go:1]: https://github.com/onflow/flow-go/pull/4857/commits/1acc984c87e444a5e2cdaf4d957ff12e74f1de71

File: network/p2p/scoring/registry.go [hunk 3]
```diff
@@ -48,27 +48,34 @@ const (
 	iHaveMisbehaviourPenalty = -10
 	// iWantMisbehaviourPenalty is the penalty applied to the application specific penalty when a peer conducts a iWant misbehaviour.
 	iWantMisbehaviourPenalty = -10
+	// clusterPrefixedPenaltyReductionFactor factor used to reduce the penalty for control message misbehaviours on cluster prefixed topics. This is allows a more lenient punishment for nodes
```
Comment (gomisha) [inline network/p2p/scoring/registry.go:51]:
```suggestion
	// clusterPrefixedPenaltyReductionFactor factor used to reduce the penalty for control message misbehaviours on cluster prefixed topics. This allows a more lenient punishment for nodes
```

File: network/p2p/scoring/registry_test.go [hunk 4]
```diff
@@ -422,6 +423,63 @@ func TestPersistingInvalidSubscriptionPenalty(t *testing.T) {
...
+	// the penalty should now be updated in the spamRecords
+	record, err, ok := spamRecords.Get(peerID) // get the record from the spamRecords.
+	assert.True(t, ok)
+	assert.NoError(t, err)
+	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-3)
+	assert.Equal(t, scoring.InitAppScoreRecordState().Decay, record.Decay)
+	// this peer has a spam record, with no subscription penalty. Hence, the app specific score should only be the spam penalty,
+	// and the peer should be deprived of the default reward for its valid staked role.
+	score = reg.AppSpecificScoreFunc()(peerID)
+	assert.Less(t, math.Abs(expectedPenalty-score), 10e-3)
```
Comment (gomisha) [inline network/p2p/scoring/registry_test.go:480]: How did you choose the tolerance for `assert.Less`? I'm seeing this test pass for 2 orders of magnitude smaller:

```
	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-5)
	assert.Equal(t, scoring.InitAppScoreRecordState().Decay, record.Decay)
	// this peer has a spam record, with no subscription penalty. Hence, the app specific score should only be the spam penalty,
	// and the peer should be deprived of the default reward for its valid staked role.
	score = reg.AppSpecificScoreFunc()(peerID)
	assert.Less(t, math.Abs(expectedPenalty-score), 10e-5)
```

**Review by gomisha — APPROVED** [review]
A couple of nits but looks good overall.

File: network/p2p/inspector/validation/control_message_validation_inspector.go [hunk 5]
```diff
@@ -687,23 +694,23 @@ func (c *ControlMsgValidationInspector) performSample(ctrlMsg p2pmsg.ControlMess
 //
 // This func returns an exception in case of unexpected bug or state corruption if cluster prefixed topic validation
 // fails due to unexpected error returned when getting the active cluster IDS.
-func (c *ControlMsgValidationInspector) validateTopic(from peer.ID, topic channels.Topic, activeClusterIds flow.ChainIDList) error {
+func (c *ControlMsgValidationInspector) validateTopic(from peer.ID, topic channels.Topic, activeClusterIds flow.ChainIDList) (error, bool) {
```
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:697]:
```suggestion
func (c *ControlMsgValidationInspector) validateTopic(from peer.ID, topic channels.Topic, activeClusterIds flow.ChainIDList) (error, p2p.CtrlMsgTopicType) {
```

File: network/p2p/inspector/validation/control_message_validation_inspector.go [hunk 6]
```diff
@@ -687,23 +694,23 @@ func (c *ControlMsgValidationInspector) performSample(ctrlMsg p2pmsg.ControlMess
...
 	channel, ok := channels.ChannelFromTopic(topic)
 	if !ok {
-		return channels.NewInvalidTopicErr(topic, fmt.Errorf("failed to get channel from topic"))
+		return channels.NewInvalidTopicErr(topic, fmt.Errorf("failed to get channel from topic")), false
```
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:700]:
```suggestion
		return channels.NewInvalidTopicErr(topic, fmt.Errorf("failed to get channel from topic")), p2p.CtrlMsgNonClusterTopicType
```

File: network/p2p/inspector/validation/control_message_validation_inspector.go [hunk 7]
```diff
@@ -687,23 +694,23 @@ func (c *ControlMsgValidationInspector) performSample(ctrlMsg p2pmsg.ControlMess
...
 	}
 
 	// handle cluster prefixed topics
 	if channels.IsClusterChannel(channel) {
-		return c.validateClusterPrefixedTopic(from, topic, activeClusterIds)
+		return c.validateClusterPrefixedTopic(from, topic, activeClusterIds), true
```
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:705]:
```suggestion
		return c.validateClusterPrefixedTopic(from, topic, activeClusterIds), p2p.CtrlMsgTopicTypeClusterPrefixed
```

File: network/p2p/inspector/validation/control_message_validation_inspector.go [hunk 8]
```diff
@@ -687,23 +694,23 @@ func (c *ControlMsgValidationInspector) performSample(ctrlMsg p2pmsg.ControlMess
...
 	}
 
 	// non cluster prefixed topic validation
 	err := channels.IsValidNonClusterFlowTopic(topic, c.sporkID)
 	if err != nil {
-		return err
+		return err, false
```
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:711]:
```suggestion
		return err, p2p.CtrlMsgNonClusterTopicType
```

File: network/p2p/inspector/validation/control_message_validation_inspector.go [hunk 9]
```diff
@@ -687,23 +694,23 @@ func (c *ControlMsgValidationInspector) performSample(ctrlMsg p2pmsg.ControlMess
...
 	}
-	return nil
+	return nil, false
```
Comment (yhassanzadeh13) [inline network/p2p/inspector/validation/control_message_validation_inspector.go:713]:
```suggestion
	return nil, p2p.CtrlMsgTopicTypeClusterPrefixed
```

File: network/p2p/scoring/registry_test.go (hunk 4 above)
Comment (yhassanzadeh13) [inline network/p2p/scoring/registry_test.go:480]:
```suggestion
	tolerance := 10e-3 // 0.1%
	if expectedPenalty == 0 {
		assert.Less(t, math.Abs(expectedPenalty), tolerance)
	} else {
		assert.Less(t, math.Abs(expectedPenalty-score)/expectedPenalty, tolerance)
	}
```

File: network/p2p/scoring/registry_test.go [hunk 10]
```diff
@@ -422,6 +423,63 @@ func TestPersistingInvalidSubscriptionPenalty(t *testing.T) {
 	assert.Equal(t, 0.0, record.Penalty) // penalty should be zero.
 }
 
+// TestPeerSpamPenaltyClusterPrefixed evaluates the application-specific penalty calculation for a node when a spam record is present
+// for cluster-prefixed topics. In the case of an invalid control message notification marked as cluster-prefixed,
+// the application-specific penalty should be reduced by the default reduction factor. This test verifies the accurate computation
+// of the application-specific score under these conditions.
+func TestPeerSpamPenaltyClusterPrefixed(t *testing.T) {
```
Comment (yhassanzadeh13) [inline network/p2p/scoring/registry_test.go:430]: This test only concerns `GRFAT` control message types; please extend it to support other message types as well in the same scenario.

**Review by yhassanzadeh13 — APPROVED** [review]

File: network/p2p/scoring/registry_test.go (hunk 4 above)
Comment (kc1116) [reply network/p2p/scoring/registry_test.go:480]: We are compensating for the potential difference in resources between local machine and remote CI instances. Also the difference is very small and does not decrease or increase the effectiveness of the test.

File: network/p2p/scoring/registry_test.go (hunk 10 above)
Comment (kc1116) [reply network/p2p/scoring/registry_test.go:430]: https://github.com/onflow/flow-go/pull/4857/commits/1749d789e61dff2384e7ad8ef332a9bda642bd01

**Pull request merged**